pkg/cloudprovider/providers/openstack/openstack_volumes.go

```diff
@@ -317,8 +319,33 @@ func (os *OpenStack) AttachDisk(instanceID, volumeID string) (string, error) {
 		if instanceID == volume.AttachedServerId {
 			glog.V(4).Infof("Disk %s is already attached to instance %s", volumeID, instanceID)
 			return volume.ID, nil
+		} else {
+			nodeName, err := os.GetNodeNameByID(volume.AttachedServerId)
+			attachErr := fmt.Sprintf("disk %s path %s is attached to a different instance (%s)", volumeID, volume.AttachedDevice, volume.AttachedServerId)
+			if err != nil {
+				glog.Error(attachErr)
+				return "", errors.New(attachErr)
+			}
+			// using volume.AttachedDevice may cause problems because cinder does not report device path correctly see issue #33128
+			devicePath := volume.AttachedDevice
+			danglingErr := volumeutil.NewDanglingError(attachErr, nodeName, devicePath)
+			glog.V(4).Infof("volume %s is already attached to node %s path %s", volumeID, nodeName, devicePath)
+			// check special case, if node is deleted from cluster but exist still in openstack
+			// we need to check can we detach the cinder, node is deleted from cluster if state is not ACTIVE
+			srv, err := getServerByName(cClient, nodeName, false)
+			if err != nil {
+				return "", err
+			}
+			if srv.Status != "ACTIVE" {
+				err = os.DetachDisk(volume.AttachedServerId, volumeID)
```

Comment:
Why are we performing detach in `attach` function? I thought - we made it clear that raising `NewDanglingError` should suffice